### PR TITLE
Pure Texture to Sampled Texture Transform

### PR DESCRIPTION
**Pull request by TiemoJung**
Adds a transformation step at the post processing step.
Two modes are available:
1) keep
- Keeps samplers, textures and sampled textures as is
2) transform pure texture into sampled texture and remove pure samplers
- removes all pure samplers
- transforms all pure textures into its sampled counter part

replacement for #674

### CODE REVIEW

**Comment by hrydgard** [comment]
Sorry for the bump, but any plans to merge this? Would be useful.

Needs a rebase by the way.

File: glslang/MachineIndependent/Intermediate.cpp [hunk 1]
```diff
@@ -2854,4 +2862,41 @@ bool TIntermediate::specConstantPropagates(const TIntermTyped& node1, const TInt
            (node2.getType().getQualifier().isSpecConstant() && node1.getType().getQualifier().isConstant());
 }
 
+struct TextureUpgradeAndSamplerRemovalTransform : public TIntermTraverser {
+    bool visitAggregate(TVisit, TIntermAggregate* ag) override {
+        using namespace std;
+        TIntermSequence& seq = ag->getSequence();
+        // remove pure sampler variables
+        TIntermSequence::iterator newEnd = remove_if(seq.begin(), seq.end(), [=](TIntermNode* node) {
```
Comment (johnkslang) [inline glslang/MachineIndependent/Intermediate.cpp]: What is it you are trying to capture by value here?  Would `[&]` or even `[ ]` be better?

File: glslang/MachineIndependent/Intermediate.cpp [hunk 2]
```diff
@@ -2854,4 +2862,41 @@ bool TIntermediate::specConstantPropagates(const TIntermTyped& node1, const TInt
...
+            TIntermSymbol* symbol = node->getAsSymbolNode();
+            if (!symbol)
+                return false;
+
+            return (symbol->getBasicType() == EbtSampler && symbol->getType().getSampler().isPureSampler());
+        });
+        seq.erase(newEnd, seq.end());
+        // replace constructors with sampler/textures
+        // update textures into sampled textures
+        for_each(seq.begin(), seq.end(), [=](TIntermNode*& node) {
```
Comment (johnkslang) [inline glslang/MachineIndependent/Intermediate.cpp]: Same question; unclear why to request capturing everything by value.

**Comment by johnkslang** [comment]
If this is useful, it seems reasonable. Can merge in when it is ready (and question answered).

**Comment by TiemoJung** [comment]
@johnkslang The two lambda captures where not needed, and i removed it. Also rebased it.

**Pull request merged**